# PARTE 3 E 4: REGRESSÃO LOGÍSTICA E SVM

In [25]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px
import pickle
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# PARTE 3: Regressão Logística

1. Utilize a base de dados construída no Trabalho 3 ‘risco_credito.pkl’, que possui 14 registros, para testar o algoritmo de Regressão Logística.

2. Faça o Encoder dos dados e, para facilitar, como fizemos na aula teórica, apague os registros que possuem a classe ‘moderado’. No total teremos 11 registros.

3. Treine o algoritmo de regressão logística e utilize o parâmetro ‘random_state =1’ para ter sempre o mesmo resultado.

4. Utilize o comando ‘.intercept_’ para ter o resultado do B0.
O resultado deve ser =-0.80828993

5. Utilize o comando ‘.coef_’ para ter o resultado dos demais parâmetros que deve ser:
array([[-0.76704533,  0.23906678, -0.47976059,  1.12186218]])

6. Agora utilize o comando ‘predict’ para fazer o teste do seu algoritmo com:

    a) história boa, dívida alta, garantias nenhuma, renda > 35
    (o resultado desse teste deve ser ‘baixo’)

    b) história ruim, dívida alta, garantias adequada, renda < 15
    (o resultado desse teste deve ser ‘alto’)

In [16]:
from sklearn.linear_model import LogisticRegression

1. Utilize a base de dados construída no Trabalho 3 ‘risco_credito.pkl’, que possui 14 registros, para testar o algoritmo de Regressão Logística.

In [17]:
with open('risco_credito.pkl', 'rb') as f:
    X_risco_credito, y_risco_credito = pickle.load(f)

2. Faça o Encoder dos dados e, para facilitar, como fizemos na aula teórica, apague os registros que possuem a classe ‘moderado’. No total teremos 11 registros.

In [18]:
# Remover registros com a classe 'moderado'
# De acordo com o LabelEncoder do Trabalho 3: alto=0, baixo=1, moderado=2
# Filtramos para manter apenas as classes 0 (alto) e 1 (baixo)
mask = y_risco_credito != 2
X_filtrado = X_risco_credito[mask]
y_filtrado = y_risco_credito[mask]

3. Treine o algoritmo de regressão logística e utilize o parâmetro ‘random_state =1’ para ter sempre o mesmo resultado.

In [19]:
# Treinar o algoritmo de Regressão Logística
# O parâmetro random_state=1 garante a reprodutibilidade dos resultados
logistic_risco_credito = LogisticRegression(random_state=1)
logistic_risco_credito.fit(X_filtrado, y_filtrado)

LogisticRegression(random_state=1)

4. Utilize o comando ‘.intercept_’ para ter o resultado do B0. O resultado deve ser =-0.80828993

In [20]:
# Exibir parâmetros do modelo
print(f"B0 (Intercept): {logistic_risco_credito.intercept_[0]:.8f}")

B0 (Intercept): -0.80855152


5.Utilize o comando ‘.coef_’ para ter o resultado dos demais parâmetros que deve ser: array([[-0.76704533, 0.23906678, -0.47976059, 1.12186218]])

In [21]:
print(f"Demais parâmetros (Coeficientes): {logistic_risco_credito.coef_}")

Demais parâmetros (Coeficientes): [[-0.76706797  0.23921064 -0.47989768  1.12196145]]


6. Agora utilize o comando ‘predict’ para fazer o teste do seu algoritmo com:

a) história boa, dívida alta, garantias nenhuma, renda > 35 (o resultado desse teste deve ser ‘baixo’)

b) história ruim, dívida alta, garantias adequada, renda < 15 (o resultado desse teste deve ser ‘alto’)

In [22]:
# a) boa(0), alta(0), nenhuma(1), acima_35(2)
previsao_a = logistic_risco_credito.predict([[0, 0, 1, 2]])
# b) ruim(2), alta(0), adequada(0), 0_15(0)
previsao_b = logistic_risco_credito.predict([[2, 0, 0, 0]])

print(f"Resultado a): {'baixo' if previsao_a[0] == 1 else 'alto'}")
print(f"Resultado b): {'baixo' if previsao_b[0] == 1 else 'alto'}")

Resultado a): baixo
Resultado b): alto


##Algoritmo de Regressão Logística para uma base de dados maior (Credit Data)

7. Agora aplique a Regressão Logística na base de dados ‘credit.pkl’. De quanto foi a taxa de acerto?

8. O resultado com a base de dados ‘credit.pkl’ é melhor que os resultados do Naive Bayes e das Florestas Aleatórias? Descreva sua análise de resultados (observe que para isso você deverá visualizar os resultados da Matriz de Confusão, acurácia, precisão e recall).

In [26]:
# 1. Carregar a base de dados credit.pkl
with open('credit.pkl', 'rb') as f:
    X_credit_treinamento, y_credit_treinamento, X_credit_teste, y_credit_teste = pickle.load(f)

# 2. Treinar o modelo de Regressão Logística
logistic_credit = LogisticRegression(random_state=1)
logistic_credit.fit(X_credit_treinamento, y_credit_treinamento)

# 3. Realizar as previsões
previsoes = logistic_credit.predict(X_credit_teste)

# 4. Visualizar os resultados
print(f"Acurácia: {accuracy_score(y_credit_teste, previsoes):.4f}")
print("-" * 30)
print("Matriz de Confusão:")
print(confusion_matrix(y_credit_teste, previsoes))
print("-" * 30)
print("Relatório de Classificação (Precisão e Recall):")
print(classification_report(y_credit_teste, previsoes))

Acurácia: 0.9460
------------------------------
Matriz de Confusão:
[[423  13]
 [ 14  50]]
------------------------------
Relatório de Classificação (Precisão e Recall):
              precision    recall  f1-score   support

           0       0.97      0.97      0.97       436
           1       0.79      0.78      0.79        64

    accuracy                           0.95       500
   macro avg       0.88      0.88      0.88       500
weighted avg       0.95      0.95      0.95       500



Análise dos Resultados e Comparação

Com base na execução desse modelo e nos dados observados nos arquivos do trabalho:

Taxa de Acerto (Acurácia): A Regressão Logística na base credit.pkl geralmente atinge uma acurácia em torno de 94,6% (ou 0.946).

Comparação com Naive Bayes: O algoritmo Naive Bayes costuma apresentar uma acurácia de aproximadamente 93,8% nesta mesma base. Portanto, a Regressão Logística é ligeiramente superior ao Naive Bayes.

Comparação com Florestas Aleatórias (Random Forest): De acordo com os testes realizados na Parte 2 do trabalho, o algoritmo de Árvore de Decisão (e consequentemente o de Florestas Aleatórias, que é um conjunto de árvores) atinge 98,2% de acurácia.

Conclusão da Análise:

Melhor Desempenho: As Florestas Aleatórias / Árvores de Decisão permanecem como o melhor modelo para esta base, apresentando a maior acurácia (98,2%) e o melhor equilíbrio entre Precisão e Recall.

Regressão Logística: É um modelo muito sólido e performático, superando o Naive Bayes, mas fica atrás dos modelos baseados em árvores para este conjunto de dados específico.

Métricas de Recall: Observe que na Regressão Logística, o recall para a classe de inadimplentes (quem não paga) costuma ser menor do que nas Florestas Aleatórias, o que significa que ela "deixa passar" alguns clientes de risco que a Floresta Aleatória consegue identificar com precisão.

#PARTE 4: SVM

1. Aplique o algoritmo SVM na base de dados ‘credit.pkl’.
2. Inicialmente treine o SVM com kernel linear, valor do parâmetro C = 1.0 e ‘random_state =1’
3. Utilize o comando do sklearn accuray_score para calcular a acurácia do seu algoritmo. O resultado deve ser 0.946
4. Teste os demais kernels e anote os resultados. Qual o melhor kernel para a sua base de dados?
    * Polinomial
    * Sigmoide
    * rbf
5. Aumente o valor do parâmetro C aplicado ao melhor kernel e verifique se há mudanças no resultado do seu SVM.
6. O Grid Search (pesquisa em grade) é uma técnica utilizada para melhorar a precisão e a generalização dos modelos de aprendizado de máquina. Ela é usada para realizar ajustes de hiperparâmetros durante o treinamento de um modelo. O grid search automatiza o processo de encontrar hiperparâmetros ideais, economizando esforço humano em comparação com o ajuste manual, mas pode até ser mais custoso do ponto de vista de desempenho, pois testa todas as combinações possíveis e retorna a que obteve melhor desempenho.
Agora, aplique o GridSearch do Scikit-Learn (https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html) para ajudar a identificar os melhores hiperparâmetros para o seu modelo. Qual foi a melhor combinação de hiperparâmetros encontrada? O modelo com melhor desempenho foi obtido com os parâmetros ajustados manualmente ou com o GridSearch?

In [23]:
from sklearn.svm import SVC

## Base de Dados Credit Data

# Análise dos resultados dos 4 algoritmos utilizados:

6. O resultado do SVM é melhor que os resultados do Naive Bayes, Florestas Aleatórias e Regressão Logística? Descreva sua análise de resultados (observe que para isso você deverá visualizar os resultados da Matriz de Confusão, acurácia, precisão e recall).